### DISCUSSION TITLE
Make the Symfony container available for front end environment.

### DISCUSSION THREAD

**Discussion by FabienPapet** [PrestaShop/PrestaShop#30867] [Ideas]
# Description 

Instead of calling the legacy with Symfony, we will be able to call Symfony instead. This will allow a better service reusability: Services can be defined and registered with the container, making them available for use throughout the application. This can help to reduce code duplication and improve maintainability.

This will be needed for the cart rules refactoring as many code is used on front and back office.

This is **not** about having Symfony controllers for the front end.

There's already a PR: https://github.com/PrestaShop/PrestaShop/pull/29997

# Release date ?
Next major version of PS (9.0.0)

# Todo

Steps:
 - [x] Create a Symfony container class that will give access to the container (do not rely on the adapter class as it relies on the `global $kernel;` variable.
- [x]  for starters the "front services" initialized in these files https://github.com/PrestaShop/PrestaShop/tree/47d2daba6e140c36d2894092d8c4a3de0c39ab3c/config/services/front won't be loaded Thus some services will not be accessible which is a breaking change
- [ ]  this can be done thanks to our compiler pass like it's been done here for admin services https://github.com/PrestaShop/PrestaShop/blob/ff96226eae149d58b34ca6c5228b846595d05914/src/PrestaShopBundle/PrestaShopBundle.php#L64
- [ ]  Replace the current custom Container with the symfony one, it means the reference in Context and Controller classes must be correctly filled:
  - for front controller https://github.com/PrestaShop/PrestaShop/blob/dc2c310af81a1c1a2b648462fd3007d3fc88e97e/classes/controller/FrontController.php#L2051
  - for context it's the only write access I could find so far (which is worrying ^^) https://github.com/PrestaShop/PrestaShop/blob/c13af1ac327d14f42e5a505eb3b46d3adc534eec/classes/Tools.php#L787

- [ ]  Deprecate our `PrestaShop\PrestaShop\Adapter\ContainerBuilder` service and make sure it's not used anywhere anymore
- [ ]  Check that the previous job of the legacy container is still covered by the new one (for example doctrine entities from modules, productcomments is a good module to test this)
- [ ] The legacy `ServiceLocator` should be based on the SF container, then we can replace all its usages and finally deprecate it
- [ ] Install process should be based on the SF container
- [ ] The container should be bootable without the shop being installed (including in CLI)
- [ ]  Make some performance checks using Blackfire

# To be discussed 

> - We'll need a new parameter for our AppKernel, similar to environment (prod, dev, test) and debugMode (tru, false) we'll need a third one for front/admin (I don't know how to name it) This will allow us relying on the same kernel/container but we'll be able to adapt some services/configurations/routing differently depending on the provided environment

That means you are coupling your DI with the request context to decide which services you can use or not. I'm 100% against it introduces too much complexity. If we want to customize things, there are plenty of other options like route prefixes, listeners, attributes. Split Frontend and admin bundle etc. Adding this behaviour is re-creating the frontend container we just want to remove. I can give you examples of actual PRs that fails because we edit a service and add a dependency that is not defined in the frontend. It is a nightmare

**Comment by matks** [comment]
That's a lovely idea ❤️

**Comment by eternoendless** [comment]
> Install process should be based on the SF container

I strongly advise not to share the same code between install/update and normal operations. We have already had problems in the past due to services expecting dependencies or assets to be present and ready, but weren't so during the install/update process.

We could make the installer an independent SF app, though

**Comment by Dreimus** [comment]
Actually I use it on a project, works kinda properly.

Watch out for the order process because the StockMvt ORM class requires an id_employee ( https://github.com/PrestaShop/PrestaShop/blob/78ca7ec56304f76b7ca6aefb7b009c7319258580/src/PrestaShopBundle/Entity/StockMvt.php#L80 )

And the StockManager evaluate SymfonyContainer instanciation to use the ORM class ( https://github.com/PrestaShop/PrestaShop/blob/78ca7ec56304f76b7ca6aefb7b009c7319258580/src/Core/Stock/StockManager.php#L362 )

So, if a customer place an order, that will induce a stock movement, then the ordering process crash, because there is no employee.
